src/buffer/out/Row.cpp

```diff
@@ -191,8 +347,331 @@ OutputCellIterator ROW::WriteCells(OutputCellIterator it, const til::CoordType i
+    for (; colExtBeg != 0 && _uncheckedIsTrailer(colExtBeg); --colExtBeg)
+    {
+    }
+
+    // Extend range upwards (trailing whitespace)
+    uint16_t colExtEnd = colEnd;
+    // Safety: colExtEnd cannot be incremented past _columnCount, because the last
+    // _charOffset at index _columnCount will never get the CharOffsetsTrailer flag.
+    for (; _uncheckedIsTrailer(colExtEnd); ++colExtEnd)
+    {
+    }
+    // Safety: After the previous loop colExtEnd is [0, _columnCount].
+    const uint16_t chExtEnd = _uncheckedCharOffset(colExtEnd);
+
+    const uint16_t leadingSpaces = colBeg - colExtBeg;
+    const uint16_t trailingSpaces = colExtEnd - colEnd;
+    const size_t chExtEndNew = chars.size() + leadingSpaces + trailingSpaces + chExtBeg;
+
+    if (chExtEndNew != chExtEnd)
+    {
+        _resizeChars(colExtEnd, chExtBeg, chExtEnd, chExtEndNew);
+    }
+
+    // Add leading/trailing whitespace and copy chars
+    {
+        auto it = _chars.begin() + chExtBeg;
+        it = fill_n_small(it, leadingSpaces, L' ');
+        it = copy_n_small(chars.begin(), chars.size(), it);
+        it = fill_n_small(it, trailingSpaces, L' ');
+    }
+    // Update char offsets with leading/trailing whitespace and the chars columns.
+    {
+        auto chPos = chExtBeg;
+        auto it = _charOffsets.begin() + colExtBeg;
+
+        it = iota_n_mut(it, leadingSpaces, chPos);
+
+        *it++ = chPos;
+        it = fill_small(it, _charOffsets.begin() + colEnd, gsl::narrow_cast<uint16_t>(chPos | CharOffsetsTrailer));
+        chPos = gsl::narrow_cast<uint16_t>(chPos + chars.size());
+
+        it = iota_n_mut(it, trailingSpaces, chPos);
+    }
+}
+
+const til::small_rle<TextAttribute, uint16_t, 1>& ROW::Attributes() const noexcept
+{
+    return _attr;
+}
+
+TextAttribute ROW::GetAttrByColumn(const til::CoordType column) const
+{
+    return _attr.at(_clampedUint16(column));
+}
+
+std::vector<uint16_t> ROW::GetHyperlinks() const
+{
+    std::vector<uint16_t> ids;
+    for (const auto& run : _attr.runs())
+    {
+        if (run.value.IsHyperlink())
+        {
+            ids.emplace_back(run.value.GetHyperlinkId());
+        }
+    }
+    return ids;
+}
+
+uint16_t ROW::size() const noexcept
+{
+    return _columnCount;
+}
+
+til::CoordType ROW::MeasureLeft() const noexcept
+{
+    const auto text = GetText();
+    const auto beg = text.begin();
+    const auto end = text.end();
+    auto it = beg;
+
+    for (; it != end; ++it)
+    {
+        if (*it != L' ')
+        {
+            break;
+        }
+    }
+
+    return gsl::narrow_cast<til::CoordType>(it - beg);
+}
+
+til::CoordType ROW::MeasureRight() const noexcept
+{
+    const auto text = GetText();
+    const auto beg = text.begin();
+    const auto end = text.end();
+    auto it = end;
+
+    for (; it != beg; --it)
+    {
+        // it[-1] is safe as `it` is always greater than `beg` (loop invariant).
+        if (til::at(it, -1) != L' ')

```

Comment:
Why is this cheaper than `std::find_last_of` or `std::find_first_not_of(rbegin...`? Or, is it not?

Reply:
You mean `std::basic_string_view::find_last_not_of`? It's not cheaper, but it's easier to follow this C-ish code IMO, because `find_last_not_of` isn't easy to use here:
```cpp
til::CoordType ROW::MeasureRight() const noexcept
{
    const auto text = GetText();
    const auto idx = text.find_last_not_of(L' ');
    const auto skipped = idx == std::wstring_view::npos ? text.size() : text.size() - idx - 1;
    return gsl::narrow_cast<til::CoordType>(_columnCount - skipped);
}
```